_posts/2022-09-08-patterns-with-suspense.markdown

```diff
@@ -0,0 +1,594 @@
+</Suspense>
+
+<Autocomplete ... />
+```
+
+Full example:
+
+```tsx
+const ArtistAutocomplete = () => {
+  const [query, setQuery] = useState(null)
+  const [options, setOptions] = useState([])
+
+  const handleOnChange = (response) => {
+    setOptions(response)
+  }
+
+  return (
+    <>
+      <Suspense fallback="Loading...">
+        <Fetcher query={query} onChange={handleOnChange} />
+      </Suspense>
+
+      <Autocomplete
+        onChange={setQuery}
+        options={options}
+        isLoading="???" // where do we get loading state?
+      />
+    </>
+  )
+}
+
+const Fetcher = ({ query, onChange }) => {
+  const response = useArtistQuery(query) // triggers suspense boundary
+
+  useEffect(() => {
+    onChange(response)
+  }, [response])
+
+  return null
+}
+```
+
+The `Fetcher` component doesn't actually render anything; it just fetches data
+via the `useArtistQuery` hook -- and hooks, as you know, can **only** be used in
+React components. Once the data has been returned it sends the response back to
+the parent via the `onChange` callback prop. This then pipes the data into our
+`<Autocomplete />` component.
+
+With our flickering UI problem now solved, we're left with figuring out how to
+update the `isLoading` prop on our `<Autocomplete />` component. How? We need to
+create a loading toggle component that is updated every time a component
+suspends and pass that in as the suspense fallback:
+
+```tsx
+<Suspense fallback={<LoadingToggle />}>
+  <Fetcher />
+</Suspense>
+
+<Autocomplete ... />
+```
+
+Full example:
+
+```tsx
+const ArtistAutocomplete = () => {
+  const [query, setQuery] = useState(null)
+  const [options, setOptions] = useState([])
+  const [isLoading, setIsLoading] = useState(false)
+
+  const handleOnChange = (response) => {
+    setIsLoading(false)
+    setOptions(response)
+  }
+
+  return (
+    <>
+      <Suspense fallback={<LoadingToggle onChange={setIsLoading} />}>
+        <Fetcher query={query} onChange={handleResponse} />
+      </Suspense>
+
+      <Autocomplete
+        onChange={setQuery}
+        options={options}
+        isLoading={isLoading}
+      />
+    </>
+  )
+}
+
+const Fetcher = ({ query, onChange }) => {
+  const response = useArtistQuery(query)
+
+  useEffect(() => {
+    onChange(response)
+  }, [response])
+
+  return null
+}
+
+const LoadingToggle = ({ onChange }) => {
+  useEffect(() => {
+    onChange(true)
+  }, [])
+
+  return null
+}
+```
+
+When the suspense boundary fallback has been triggered the `<LoadingToggle />`
+component renders, and instantly we call the `onChange` callback with `true` to
+alert the parent that our component has suspended. And again, similar to
+`<Fetcher />`, we return `null` from the `LoadingToggle` because all we're
+concerned about is learning when the suspense boundary has been triggered to set
+the state in the parent, and we don't need to render visuals for the fallback;
+our aim is solely to update the `isLoading` prop on `<Autocomplete />`.
```

Comment:
Got a bit lost reading this bit, maybe we could break it down:

```suggestion
alert the parent that our component has suspended. And again, similar to
`<Fetcher />`, we return `null` from the `LoadingToggle`. That's because all we're
concerned about is learning when the suspense boundary has been triggered: once
we know it's been triggered, we set the state in the parent. And we don't need to 
render visuals for the fallback; our aim is solely to update the `isLoading` prop on
`<Autocomplete />`.
```